test/coupler_compatibility.jl

```diff
@@ -0,0 +1,171 @@
+# its own surface fields, and then it copies the data from those fields into the
+# cache by calling set_surface_conditions!.
+
+# Since overwriting the internal data structures of ClimaAtmos makes it harder
+# to develop a modular API, the pattern demonstrated in the first test should
+# only be used to quickly write tests for the coupler. For more long-term code,
+# the pattern demonstrated in the second test is preferable.
+
+@testset "Coupler Compatibility (Hacky Version)" begin
+    # Initialize a model. The value of --surface_setup is irrelevant, since it
+    # will get overwritten.
+    config = CA.AtmosConfigArgs(;
+        args = ["--initial_condition", "DryBaroclinicWave"],
+    )
+    integrator = CA.get_integrator(config)
+    (; p, t) = integrator
+    Y = integrator.u
+    FT = eltype(Y)
+    thermo_params = CAP.thermodynamics_params(p.params)
+
+    # Override p.sfc_setup with a Field of SurfaceStates. The value of T is
+    # irrelevant, since it will get updated.
+    surface_state = CA.SurfaceConditions.SurfaceState(;
+        parameterization = CA.SurfaceConditions.MoninObukhov(;
+            z0m = FT(z0m),
+            z0b = FT(z0b),
+        ),
+        T = FT(NaN),
+        gustiness = FT(gustiness),
+        beta = FT(beta),
+    )
+    sfc_setup = similar(Spaces.level(Y.f, half), typeof(surface_state))
+    @. sfc_setup = (surface_state,)
+    p_overwritten = merge(p, (; sfc_setup))
+
+    # Test that set_precomputed_quantities! can be used to update the surface
+    # temperature to T1 and then to T2.
+    @. sfc_setup.T = FT(T1)
+    CA.set_precomputed_quantities!(Y, p_overwritten, t)
+    sfc_T = @. TD.air_temperature(thermo_params, p.sfc_conditions.ts)
+    @test all(isequal(T1), parent(sfc_T))
+    @. sfc_setup.T = FT(T2)
+    CA.set_precomputed_quantities!(Y, p_overwritten, t)
+    sfc_T = @. TD.air_temperature(thermo_params, p.sfc_conditions.ts)
+    @test all(isequal(T2), parent(sfc_T))
+end
+
+@testset "Coupler Compatibility (Proper Version)" begin
+    # Initialize a model. Set --surface_setup to PrescribedSurface to prevent
+    # ClimaAtmos from modifying the surface conditions.
+    config = CA.AtmosConfigArgs(;
+        args = [
+            "--initial_condition",
+            "DryBaroclinicWave",
+            "--surface_setup",
+            "PrescribedSurface",
+        ],
+    )
+    integrator = CA.get_integrator(config)
+    (; p, t) = integrator
+    Y = integrator.u
+    FT = eltype(Y)
+    thermo_params = CAP.thermodynamics_params(p.params)
+
+    # Allocate fields for storing the thermodynamic state and fluxes at the
+    # surface.
+    sfc_ts = similar(Spaces.level(Y.f, half), TD.PhaseDry{FT})
+    sfc_conditions = similar(sfc_ts, SF.SurfaceFluxConditions{FT})
+
+    # Define a function for updating these fields, given the temperature at the
+    # surface and the current cache p.
+    function update_surface_fields!(sfc_ts, sfc_conditions, surface_T, p)
```

Comment:
This is more boilerplate than I expected. Does the coupler define these functions? If so, maybe we should move them into ClimaAtmos src and have the coupler call them and change this test accordingly.

Reply:
The coupler is not currently defining these functions. I am not sure what the full set of requirements will be for these functions (I know that they will need to support computing linear combinations of land and ocean surface fluxes, which we probably don't want to do on the ClimaAtmos side), so I just added a simple example to sketch out the pattern we should aim for. Maybe after a few design iterations in the coupler, we can move them into some centralized location?